README.md
# RES in Haskell

The [RES website][RES].

[RES]: https://mjn.host.cs.st-andrews.ac.uk/egyptian/res/index.html

CHANGELOG.md
# Revision history for quadres

## 0.1.0.0 -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


test/Test.hs
{-|
Module      : Main
Description : Test suite main runner.
-}
module Main
    ( main
    )
where

import qualified Hedgehog
import qualified Test.DocTest

import qualified QuadRES.ParserTest             ( tests )
import qualified QuadRES.RESTest                ( tests )

main :: IO ()
main = do
    runHedgehogTests
    runDocTests

runHedgehogTests :: IO ()
runHedgehogTests = do
    putStrLn "\n---- Running Hedgehog Tests ----"
    mapM_ Hedgehog.checkParallel hedgehogTests
    putStrLn "---- Running Hedgehog Tests ----"

hedgehogTests :: [Hedgehog.Group]
hedgehogTests = [QuadRES.ParserTest.tests, QuadRES.RESTest.tests]

runDocTests :: IO ()
runDocTests = do
    putStrLn "\n---- Running DocTests ----"
    docTests
    putStrLn "---- Finished DocTests ----"

docTests :: IO ()
docTests = Test.DocTest.doctest
    ["-isrc", "src/QuadRES/Parser.hs", "src/QuadRES/RES.hs"]


test/QuadRES/RESTest.hs
{-|
Module      : QuadRES.RESTest
Description : Tests for the QuadRES.RES module.
-}
{-# LANGUAGE OverloadedStrings #-}
module QuadRES.RESTest
  ( tests
  )
where

import qualified Hedgehog                      as H
import           Hedgehog                       ( (===) )
import qualified Hedgehog.Gen                  as Gen
import qualified Hedgehog.Range                as Range

import qualified QuadRES.RES                   as RES

tests :: H.Group
tests = H.Group "QuadRES.RES"
                [("mkRealN and realNDigits round-trip", prop_mkRealN_trip)]

prop_mkRealN_trip :: H.Property
prop_mkRealN_trip = H.property $ do
  ones       <- H.forAll $ Gen.word8 $ Range.linear 0 9
  tenths     <- H.forAll $ Gen.word8 $ Range.linear 0 9
  hundredths <- H.forAll $ Gen.word8 $ Range.linear 0 9
  let (o', t', h') = RES.realNDigits (RES.mkRealN ones tenths hundredths)
  o' === ones
  t' === tenths
  h' === hundredths


test/QuadRES/ParserTest.hs
{-|
Module      : QuadRES.ParserTest
Description : Tests for the QuadRES.Parser module.
-}
{-# LANGUAGE OverloadedStrings #-}
module QuadRES.ParserTest
  ( tests
  )
where

import qualified Hedgehog                      as H
import           Hedgehog                       ( (===) )
import qualified Hedgehog.Gen                  as Gen
import qualified Hedgehog.Range                as Range

import qualified Data.Char                      ( isDigit )
import qualified Data.Text                     as Text
import qualified Text.Megaparsec               as MP

import qualified QuadRES.Parser                as Parser
import qualified QuadRES.RES                   as RES

tests :: H.Group
tests = H.Group
  "QuadRES.Parser"
  [ ("pRealN parses shown RealN values"      , prop_pRealN_trip)
  , ("'1'-'9' are definitely non-zero digits", unit_IsNonZeroDigit_1_9)
  , ( "isNonZeroDigit matches Data.Char.isDigit with zero exclusion"
    , prop_IsNonZeroDigit
    )
  , ("'0'-'9' are definitely digits"    , unit_IsDigit_0_9)
  , ("isDigit matches Data.Char.isDigit", prop_IsDigit)
  ]

prop_pRealN_trip :: H.Property
prop_pRealN_trip = H.property $ do
  let genDigit = Gen.word8 (Range.linear 0 9)
  d1 <- H.forAll genDigit
  d2 <- H.forAll genDigit
  d3 <- H.forAll genDigit
  let realN = RES.mkRealN d1 d2 d3
  H.tripping realN (Text.pack . show) (MP.parseMaybe Parser.pRealN)

unit_IsNonZeroDigit_1_9 :: H.Property
unit_IsNonZeroDigit_1_9 =
  H.withTests 1 $ H.property $ H.assert (all Parser.isNonZeroDigit ['1' .. '9'])

prop_IsNonZeroDigit :: H.Property
prop_IsNonZeroDigit = H.property $ do
  c <- H.forAll Gen.alphaNum
  (Data.Char.isDigit c && c /= '0') === Parser.isNonZeroDigit c

unit_IsDigit_0_9 :: H.Property
unit_IsDigit_0_9 =
  H.withTests 1 $ H.property $ H.assert (all Parser.isDigit ['0' .. '9'])

prop_IsDigit :: H.Property
prop_IsDigit = H.property $ do
  c <- H.forAll Gen.alphaNum
  Data.Char.isDigit c === Parser.isDigit c



src/QuadRES/Parser.hs
{-# LANGUAGE OverloadedStrings #-}
module QuadRES.Parser where

import qualified Control.Applicative.Combinators.NonEmpty
                                               as NonEmpty
import           Control.Monad.Identity         ( Identity )
import           Data.Char                      ( ord )
import           Data.Functor                   ( ($>) )
import           Data.Maybe                     ( fromMaybe )
import qualified Data.Map.Strict               as Map
import           Data.Set                       ( Set )
import qualified Data.Set                      as Set
import           Data.Text                      ( Text )
import qualified Data.Text                     as Text
import           Data.Void                      ( Void )
import           Data.Word                      ( Word8
                                                , Word16
                                                )
import qualified Text.Megaparsec               as MP
import           Text.Megaparsec                ( (<?>)
                                                , (<|>)
                                                )

import qualified QuadRES.RES                   as RES
import qualified QuadRES.Mnemonics             as Mnemonics

type Parsec e s a = MP.ParsecT e s Identity a

type Parser a = Parsec Void Text a

-- | Parse a 'RES.NamedGlyph'.
--
-- >>> MP.parseMaybe pNamedGlyph "A1"
-- Just (NamedGlyph (GlyphIDGardiner "A" 1 Nothing) (GlyphArgs []) (Notes []) (Switches []))
--
-- >>> MP.parseMaybe pNamedGlyph "A1[shade,red,mirror]"
-- Just (NamedGlyph (GlyphIDGardiner "A" 1 Nothing) (GlyphArgs [GlyphArgShadeState Shade,GlyphArgColor Red,GlyphArgMirrorState Mirror]) (Notes []) (Switches []))
--
-- >>> MP.parseMaybe pNamedGlyph "N25[bs,tbe]"
-- Just (NamedGlyph (GlyphIDGardiner "N" 25 Nothing) (GlyphArgs [GlyphArgShadePattern (ShadePattern (SideB :| [SideS])),GlyphArgShadePattern (ShadePattern (SideT :| [SideB,SideE]))]) (Notes []) (Switches []))
--
-- >>> MP.parseMaybe pNamedGlyph "\"[\"[blue]"
-- Just (NamedGlyph (GlyphIDShortString '[') (GlyphArgs [GlyphArgColor Blue]) (Notes []) (Switches []))
--
-- >>> MP.parseMaybe pNamedGlyph "kA^\"c\"[blue]"
-- Just (NamedGlyph (GlyphIDMnemonic "kA") (GlyphArgs []) (Notes [Note "c" [Blue]]) (Switches []))
pNamedGlyph :: Parser RES.NamedGlyph
pNamedGlyph =
    RES.NamedGlyph
        <$> pGlyphID
        <*> (MP.option (RES.GlyphArgs []) pGlyphArgs <* pWhitespace)
        <*> pNotes
        <*> pSwitches

-- | Parse a 'RES.GlyphID'.
--
-- >>> MP.parseMaybe pGlyphID "G14"
-- Just (GlyphIDGardiner "G" 14 Nothing)
--
-- >>> MP.parseMaybe pGlyphID "W24a"
-- Just (GlyphIDGardiner "W" 24 (Just 'a'))
--
-- >>> MP.parseMaybe pGlyphID "open"
-- Just GlyphIDOpen
--
-- >>> MP.parseMaybe pGlyphID "close"
-- Just GlyphIDClose
--
-- >>> MP.parseMaybe pGlyphID "\"[\""
-- Just (GlyphIDShortString '[')
--
-- >>> MP.parseMaybe pGlyphID "rxyt"
-- Just (GlyphIDMnemonic "rxyt")
--
-- >>> MP.parseMaybe pGlyphID "rxytfoo"
-- Nothing
pGlyphID :: Parser RES.GlyphID
pGlyphID =
    (   RES.GlyphIDGardiner
        <$> pCategory
        <*> pNonZeroNat
        <*> MP.optional pLowerLetter
        )
        <|> (MP.chunk "open" $> RES.GlyphIDOpen)
        <|> (MP.chunk "close" $> RES.GlyphIDClose)
        <|> (RES.GlyphIDShortString <$> pShortString)
        <|> (RES.GlyphIDMnemonic <$> pMnemonic)
  where
    pCategory :: Parser Text
    pCategory =
        (Text.singleton <$> MP.satisfy (isCharInRange 'A' 'I'))
            <|> (Text.singleton <$> MP.satisfy (isCharInRange 'K' 'Z'))
            <|> MP.chunk "Aa"
            <|> MP.chunk "NL"
            <|> MP.chunk "NU"

-- | Parse a mnemonic.
--
-- The mnemonic must be in the allowed list.
--
-- >>> MP.parseMaybe pMnemonic "wDAt"
-- Just "wDAt"
--
-- >>> MP.parseMaybe pMnemonic "wDAtf"
-- Nothing
pMnemonic :: Parser Text
pMnemonic = pMnemonic' (Map.keysSet Mnemonics.mnemonics)
  where
    pMnemonic' :: Set Text -> Parser Text
    pMnemonic' nameSet = do
        candidate <- MP.takeWhile1P (Just "Mnemonic Character") isMnemonicChar
        if Set.member candidate nameSet
            then pure candidate
            else MP.failure Nothing Set.empty

-- | Parse a 'RES.GlyphArgs'; a list of glyph arguments.
--
-- >>> ga = "[rotate=90, scale=0.4]"
-- >>> MP.parseMaybe pGlyphArgs ga
-- Just (GlyphArgs [GlyphArgRotate (Rotate 90),GlyphArgScale (Scale 0.40)])
--
-- >>> MP.parseMaybe pGlyphArgs ""
-- Nothing
pGlyphArgs :: Parser RES.GlyphArgs
pGlyphArgs = RES.GlyphArgs <$> pBracketedList pGlyphArg

-- | Parse a single 'RES.GlyphArg'.
--
-- >>> MP.parseMaybe pGlyphArg "ts"
-- Just (GlyphArgShadePattern (ShadePattern (SideT :| [SideS])))
pGlyphArg :: Parser RES.GlyphArg
pGlyphArg =
    (RES.GlyphArgMirrorState <$> pMirrorState)
        <|> (RES.GlyphArgRotate <$> pRotate)
        <|> (RES.GlyphArgScale <$> pScale)
        <|> (RES.GlyphArgScaleX <$> pScaleX)
        <|> (RES.GlyphArgScaleY <$> pScaleY)
        <|> (RES.GlyphArgColor <$> pColor)
        <|> (RES.GlyphArgShadeState <$> pShadeState)
        <|> (RES.GlyphArgShadePattern <$> pShadePattern)

-- | Parse a 'RES.Rotate'.
--
-- >>> MP.parseMaybe pRotate "rotate=30"
-- Just (Rotate 30)
pRotate :: Parser RES.Rotate
pRotate = RES.Rotate <$> (MP.chunk "rotate=" *> pNat)

-- | Parse a 'RES.Scale'.
--
-- >>> MP.parseMaybe pScale "scale=2.4"
-- Just (Scale 2.40)
pScale :: Parser RES.Scale
pScale = RES.Scale <$> (MP.chunk "scale=" *> pNonZeroRealN)

-- | Parse a 'RES.ScaleX'.
--
-- >>> MP.parseMaybe pScaleX "xscale=1.1"
-- Just (ScaleX 1.10)
pScaleX :: Parser RES.ScaleX
pScaleX = RES.ScaleX <$> (MP.chunk "xscale=" *> pNonZeroRealN)

-- | Parse a 'RES.ScaleY'.
--
-- >>> MP.parseMaybe pScaleY "yscale=5"
-- Just (ScaleY 5.00)
pScaleY :: Parser RES.ScaleY
pScaleY = RES.ScaleY <$> (MP.chunk "yscale=" *> pNonZeroRealN)

-- | Parse a 'RES.Notes', which is multiple notes.
--
-- >>> MP.parseMaybe pNotes "^\"a\" ^\"b\""
-- Just (Notes [Note "a" [],Note "b" []])
pNotes :: Parser RES.Notes
pNotes = RES.Notes <$> MP.many pNote

-- | Parse a 'RES.Note'.
--
-- >>> MP.parseMaybe pNote "^\"This is a note\"[blue]  "
-- Just (Note "This is a note" [Blue])
pNote :: Parser RES.Note
pNote =
    RES.Note
        <$> (MP.single '^' *> pString)
        <*> MP.option [] (pBracketedList pColor)
        <*  pWhitespace

-- | Parse whitespace and optional switches.
ws :: Parser RES.Switches
ws = pWhitespace *> pSwitches

-- | Parse a 'RES.Switches'.
--
-- >>> sw = "![red] ![noshade,sep=4.3]"
-- >>> MP.parseMaybe pSwitches sw
-- Just (Switches [Switch [SwColor Red],Switch [SwShadeState NoShade,SwSep (Sep 4.30)]])
pSwitches :: Parser RES.Switches
pSwitches = RES.Switches <$> MP.many pSwitch

-- | Parse a 'RES.Switch'.
--
-- >>> sw = "![blue, noshade, sep=1.2]"
-- >>> MP.parseMaybe pSwitch sw
-- Just (Switch [SwColor Blue,SwShadeState NoShade,SwSep (Sep 1.20)])
--
-- A plain exclamation point can also be a switch. This has no interpretation
-- other than as a possible marker to a human reader:
-- >>> MP.parseMaybe pSwitch "!"
-- Just (Switch [])
pSwitch :: Parser RES.Switch
pSwitch =
    RES.Switch
        <$> (  MP.single '!'
            *> MP.option [] (pBracketedList pSwitchArg)
            <* pWhitespace
            )

-- | Parse a 'RES.SwitchArg'.
--
-- >>> MP.parseMaybe pSwitchArg "green"
-- Just (SwColor Green)
pSwitchArg :: Parser RES.SwitchArg
pSwitchArg =
    (RES.SwColor <$> pColor)
        <|> (RES.SwShadeState <$> pShadeState)
        <|> (RES.SwSep <$> pSep)
        <|> (RES.SwFitState <$> pFitState)
        <|> (RES.SwMirrorState <$> pMirrorState)

-- | Parse a 'RES.FitState'.
--
-- >>> MP.parseMaybe pFitState "fit"
-- Just Fit
--
-- >>> MP.parseMaybe pFitState "nofit"
-- Just NoFit
pFitState :: Parser RES.FitState
pFitState =
    ((MP.chunk "fit" $> RES.Fit) <|> (MP.chunk "nofit" $> RES.NoFit))
        <?> "(fit|nofit)"

-- | Parse a 'RES.MirrorState'.
--
-- >>> MP.parseMaybe pMirrorState "mirror"
-- Just Mirror
--
-- >>> MP.parseMaybe pMirrorState "nomirror"
-- Just NoMirror
pMirrorState :: Parser RES.MirrorState
pMirrorState =
    ((MP.chunk "mirror" $> RES.Mirror) <|> (MP.chunk "nomirror" $> RES.NoMirror)
        )
        <?> "(mirror|nomirror)"

-- | Parse a 'RES.ShadeState'.
--
-- >>> MP.parseMaybe pShadeState "shade"
-- Just Shade
--
-- >>> MP.parseMaybe pShadeState "noshade"
-- Just NoShade
pShadeState :: Parser RES.ShadeState
pShadeState =
    ((MP.chunk "shade" $> RES.Shade) <|> (MP.chunk "noshade" $> RES.NoShade))
        <?> "(shade|noshade)"

-- | Parse a 'RES.Sep'.
--
-- >>> MP.parseMaybe pSep "sep=1.2"
-- Just (Sep 1.20)
--
-- >>> MP.parseMaybe pSep "sep = 1.2"
-- Nothing
pSep :: Parser RES.Sep
pSep = RES.Sep <$> (MP.chunk "sep=" *> pRealN) <?> "sep=<real>"

-- | Parse a 'RES.Color'.
--
-- >>> MP.parseMaybe pColor "blue"
-- Just Blue
pColor :: Parser RES.Color
pColor =
    (   (MP.chunk "black" $> RES.Black)
        <|> (MP.chunk "red" $> RES.Red)
        <|> (MP.chunk "green" $> RES.Green)
        <|> (MP.chunk "blue" $> RES.Blue)
        <|> (MP.chunk "white" $> RES.White)
        <|> (MP.chunk "aqua" $> RES.Aqua)
        <|> (MP.chunk "fuchsia" $> RES.Fuchsia)
        <|> (MP.chunk "gray" $> RES.Gray)
        <|> (MP.chunk "lime" $> RES.Lime)
        <|> (MP.chunk "maroon" $> RES.Maroon)
        <|> (MP.chunk "navy" $> RES.Navy)
        <|> (MP.chunk "olive" $> RES.Olive)
        <|> (MP.chunk "purple" $> RES.Purple)
        <|> (MP.chunk "silver" $> RES.Silver)
        <|> (MP.chunk "teal" $> RES.Teal)
        <|> (MP.chunk "yellow" $> RES.Yellow)
        )
        <?> "Color"

-- | Parse a 'RES.ShadePattern'.
--
-- >>> MP.parseMaybe pShadePattern "ts"
-- Just (ShadePattern (SideT :| [SideS]))
pShadePattern :: Parser RES.ShadePattern
pShadePattern = (RES.ShadePattern <$> NonEmpty.some pSide) <?> "Shade pattern"

-- | Parse a 'RES.Side' to shade.
--
-- >>> MP.parseMaybe pSide "s"
-- Just SideS
pSide :: Parser RES.Side
pSide =
    (   (MP.single 't' $> RES.SideT)
        <|> (MP.single 'b' $> RES.SideB)
        <|> (MP.single 's' $> RES.SideS)
        <|> (MP.single 'e' $> RES.SideE)
        )
        <?> "Side (t,b,s,e)"

-- | Parse a square-bracket-terminated, comma-separated list of items.
--
-- >>> MP.parseMaybe (pBracketedList pRealN) "[ ]"
-- Just []
--
-- >>> MP.parseMaybe (pBracketedList pRealN) "[ 1.3 ]"
-- Just [1.30]
--
-- >>> MP.parseMaybe (pBracketedList pRealN) "[4.3,9.5,3.2]"
-- Just [4.30,9.50,3.20]
--
-- >>> MP.parseMaybe (pBracketedList pRealN) "[ 4.3\n  , 9.5\n  , 3.2\n ]"
-- Just [4.30,9.50,3.20]
--
-- >>> MP.parseMaybe (pBracketedList pRealN) "[ 9.1, ]"
-- Nothing
--
-- >>> MP.parseMaybe (pBracketedList pRealN) "[,]"
-- Nothing
pBracketedList :: Parser a -> Parser [a]
pBracketedList pItem = do
    MP.single '[' $> ()
    pWhitespace
    items <- MP.optional (pList pItem)
    MP.single ']' $> ()
    pure (fromMaybe [] items)

-- | Parse a comma-separated, non-empty list of items.
pList :: Parser a -> Parser [a]
pList pItem = do
    pWhitespace
    item <- pItem
    pWhitespace
    optComma <- MP.optional (MP.single ',')
    case optComma of
        Nothing -> pure [item]
        Just _  -> (item :) <$> pList pItem

-- | Gobble zero or more whitespace characters.
pWhitespace :: Parser ()
pWhitespace = MP.takeWhileP (Just "whitespace") isWhitespace $> ()

---- Auxiliary Definitions

-- | Parse a natural number, in the range 0 to 999 inclusive.
--
-- >>> MP.parseMaybe pNat "1"
-- Just 1
--
-- >>> MP.parseMaybe pNat "12"
-- Just 12
--
-- >>> MP.parseMaybe pNat "123"
-- Just 123
--
-- >>> MP.parseMaybe pNat "0"
-- Just 0
--
-- >>> MP.parseMaybe pNat "999"
-- Just 999
--
-- >>> MP.parseMaybe pNat ""
-- Nothing
--
-- >>> MP.parseMaybe pNat "1000"
-- Nothing
pNat :: Parser Word16
pNat = do
    digit1    <- pDigit
    digit2opt <- MP.optional pDigit
    case digit2opt of
        Nothing     -> pure . fromIntegral $ digit1
        Just digit2 -> do
            digit3opt <- MP.optional pDigit
            case digit3opt of
                Nothing ->
                    pure $ 10 * fromIntegral digit1 + fromIntegral digit2
                Just digit3 ->
                    pure
                        $ (100 * fromIntegral digit1)
                        + (10 * fromIntegral digit2)
                        + fromIntegral digit3

-- | Parse a non-zero natural number, in the range 1 to 999 inclusive.
--
-- >>> MP.parseMaybe pNonZeroNat "1"
-- Just 1
--
-- >>> MP.parseMaybe pNonZeroNat "12"
-- Just 12
--
-- >>> MP.parseMaybe pNonZeroNat "123"
-- Just 123
--
-- >>> MP.parseMaybe pNonZeroNat "999"
-- Just 999
--
-- >>> MP.parseMaybe pNonZeroNat "0"
-- Nothing
--
-- >>> MP.parseMaybe pNonZeroNat ""
-- Nothing
--
-- >>> MP.parseMaybe pNonZeroNat "1000"
-- Nothing
pNonZeroNat :: Parser Word16
pNonZeroNat = do
    digit1    <- pNonZeroDigit
    digit2opt <- MP.optional pDigit
    case digit2opt of
        Nothing     -> pure . fromIntegral $ digit1
        Just digit2 -> do
            digit3opt <- MP.optional pDigit
            case digit3opt of
                Nothing ->
                    pure $ 10 * fromIntegral digit1 + fromIntegral digit2
                Just digit3 ->
                    pure
                        $ (100 * fromIntegral digit1)
                        + (10 * fromIntegral digit2)
                        + fromIntegral digit3

-- | Parse a real number, in the range '0.00' to '9.99'.
--
-- This accepts numbers with two decimal digits:
-- >>> MP.parseMaybe pRealN "4.32"
-- Just 4.32
--
-- We can also omit the decimal dot:
-- >>> MP.parseMaybe pRealN "6"
-- Just 6.00
--
-- The digit before the dot can be omitted:
-- >>> MP.parseMaybe pRealN ".2"
-- Just 0.20
-- >>> MP.parseMaybe pRealN ".25"
-- Just 0.25
--
-- But we cannot write a dot without any following digits:
-- >>> MP.parseMaybe pRealN "2."
-- Nothing
--
-- We also cannot include extra digits beyond those allowed:
-- >>> MP.parseMaybe pRealN "42.4"
-- Nothing
-- >>> MP.parseMaybe pRealN "2.445"
-- Nothing
--
-- We must parse something though:
-- >>> MP.parseMaybe pRealN ""
-- Nothing
pRealN :: Parser RES.RealN
pRealN = do
    digit1opt <- MP.optional pDigit
    digit2opt <- MP.optional (MP.single '.' *> pDigit)
    -- by this point, we can have obtained "x", "x.x" or ".x"; but if both
    -- digit1opt and digit2opt are empty then that's an error case
    case (digit1opt, digit2opt) of
        -- here we parsed nothing so far; so signal an error
        (Nothing, Nothing) -> MP.failure Nothing Set.empty
        -- we parsed just the first digit, but no period and following digit
        (Just d1, Nothing) -> pure (RES.mkRealN d1 0 0)
        -- we passed a period and a following digit so try for a third digit
        (_      , Just d2) -> do
            let d1 = fromMaybe 0 digit1opt
            d3 <- MP.option 0 pDigit
            pure (RES.mkRealN d1 d2 d3)

-- | Parse a non-zero real number.
pNonZeroRealN :: Parser RES.RealN
pNonZeroRealN = do
    candidate <- pRealN
    if candidate > 0 then pure candidate else MP.failure Nothing Set.empty

-- | Parses a single digit into a 'Word8'.
--
-- >>> MP.parseMaybe pDigit "4"
-- Just 4
--
-- >>> MP.parseMaybe pDigit "a"
-- Nothing
pDigit :: Parser Word8
pDigit = (\c -> fromIntegral (ord c - 48)) <$> MP.satisfy isDigit <?> "Digit"

-- | Parses a single non-zero digit into a 'Word8'.
--
-- >>> MP.parseMaybe pNonZeroDigit "4"
-- Just 4
--
-- >>> MP.parseMaybe pNonZeroDigit "0"
-- Nothing
pNonZeroDigit :: Parser Word8
pNonZeroDigit =
    (\c -> fromIntegral (ord c - 48))
        <$> MP.satisfy isNonZeroDigit
        <?> "Non-Zero Digit"

-- | Parses a single lower-case letter.
pLowerLetter :: Parser Char
pLowerLetter = MP.satisfy isLowerLetter <?> "Lower-case Letter"

-- | Parses a short string.
--
-- A short string is a string of length 1. It's equivalent to a Char in
-- Haskell, but it is delimited by double quotes in RES.
--
-- >>> MP.parseMaybe pShortString "\"Z\""
-- Just 'Z'
--
-- >>> MP.parseMaybe pShortString "\"\""
-- Nothing
--
-- >>> MP.parseMaybe pShortString "\"ab\""
-- Nothing
pShortString :: Parser Char
pShortString = (MP.single '"' *> pPrintable <* MP.single '"')
  where
    pPrintable :: Parser Char
    pPrintable =
        MP.satisfy isPrintable
            <|> (MP.chunk "\\\"" $> '"')
            <|> (MP.chunk "\\\\" $> '\\')

-- | Parses a string.
--
-- >>> MP.parseMaybe pString "\"\""
-- Just ""
--
-- >>> MP.parseMaybe pString "\"Hello World\""
-- Just "Hello World"
--
-- >>> MP.parseMaybe pString "\"Some escapes: \\\\ \\\".\""
-- Just "Some escapes: \\ \"."
--
-- >>> MP.parseMaybe pString "\""
-- Nothing
pString :: Parser Text
pString = (MP.single '"' *> pInString <* MP.single '"') <?> "String"
  where
    pInString :: Parser Text
    pInString = mconcat <$> MP.many (pPrintable <|> pEscaped)

    pPrintable :: Parser Text
    pPrintable = MP.takeWhile1P (Just "Printable Latin1 Character") isPrintable

    pEscaped :: Parser Text
    pEscaped =
        (MP.chunk "\\\"" $> Text.singleton '"')
            <|> (MP.chunk "\\\\" $> Text.singleton '\\')

---- Character Classes

-- | True if a character is a non-zero digit '1'-'9'.
--
-- >>> isNonZeroDigit '0'
-- False
--
-- >>> isNonZeroDigit '1'
-- True
isNonZeroDigit :: Char -> Bool
isNonZeroDigit c = c' >= 49 && c' <= 57
  where
    c' :: Int
    c' = ord c

-- | True if a character is a digit '0' - '9'.
--
-- >>> isDigit '4'
-- True
--
-- >>> isDigit 'a'
-- False
isDigit :: Char -> Bool
isDigit c = c' >= 48 && c' <= 57
  where
    c' :: Int
    c' = ord c

-- | True if a character is whitespace.
isWhitespace :: Char -> Bool
isWhitespace c =
    (c == ' ') || (c == '\t') || (c == '\n') || (c == '\r') || (c == '\f')

-- | True if a character can appear in a mnemonic.
isMnemonicChar :: Char -> Bool
isMnemonicChar c =
    isUpperLetter c || isLowerLetter c || (c == '1') || (c == '0')

-- | True if a character is a lower-case letter.
--
-- >>> isLowerLetter 'a'
-- True
--
-- >>> isLowerLetter 'z'
-- True
--
-- >>> isLowerLetter 'Q'
-- False
isLowerLetter :: Char -> Bool
isLowerLetter = isCharInRange 'a' 'z'

-- | True if a character is an upper-case letter.
--
-- >>> isUpperLetter 'W'
-- True
--
-- >>> isUpperLetter 'w'
-- False
isUpperLetter :: Char -> Bool
isUpperLetter = isCharInRange 'A' 'Z'

-- | True if a character is allowed as part of a string.
--
-- This excludes un-escaped quotes and backslashes.
--
-- >>> isPrintable 'a'
-- True
--
-- >>> isPrintable '"'
-- False
--
-- >>> isPrintable '\\'
-- False
isPrintable :: Char -> Bool
isPrintable c = isLatin1 c && (c /= '"') && (c /= '\\')

-- | True if a character is part of the Latin1 / ISO-8859-1 code page of the
--   ASCII table.
--
-- NOTE: This also includes double-quotes and backslash.
--
-- >>> isLatin1 '~'
-- True
--
-- >>> isLatin1 '\t'
-- False
--
-- >>> isLatin1 '"'
-- True
isLatin1 :: Char -> Bool
isLatin1 c = (c' >= 0x0020 && c' <= 0x007E) || (c' >= 0x00A0 && c' <= 0x00FF)
  where
    c' :: Int
    c' = ord c

-- | Check if a character is within a supplied inclusive range.
isCharInRange
    :: Char   -- ^ Minimum of range (inclusive).
    -> Char   -- ^ Maximum of range (inclusive).
    -> Char   -- ^ Character to check.
    -> Bool   -- ^ True if the character is in the range.
isCharInRange rangeMin rangeMax c =
    (c' >= ord rangeMin) && (c' <= ord rangeMax)
  where
    c' :: Int
    c' = ord c

-- $setup
-- >>> :set -XOverloadedStrings


src/QuadRES/RES.hs
module QuadRES.RES where

import           Data.List.NonEmpty             ( NonEmpty )
import           Data.Text                      ( Text )
import           Data.Word                      ( Word8
                                                , Word16
                                                )

-- | A named glyph.
data NamedGlyph = 
    NamedGlyph GlyphID GlyphArgs Notes Switches deriving (Eq, Show)

-- | Identifier for a glyph.
data GlyphID
    = GlyphIDGardiner Text Word16 (Maybe Char)
    | GlyphIDMnemonic Text
    | GlyphIDOpen
    | GlyphIDClose
    | GlyphIDShortString Char
    deriving (Eq, Show)

-- | List of glyph arguments.
newtype GlyphArgs = GlyphArgs [GlyphArg] deriving (Eq, Show)

-- | Single glyph argument.
data GlyphArg
    = GlyphArgMirrorState MirrorState
    | GlyphArgRotate Rotate
    | GlyphArgScale Scale
    | GlyphArgScaleX ScaleX
    | GlyphArgScaleY ScaleY
    | GlyphArgColor Color
    | GlyphArgShadeState ShadeState
    | GlyphArgShadePattern ShadePattern
    deriving (Eq, Show)

-- | Rotation of a glyph.
newtype Rotate = Rotate Word16 deriving (Eq, Show)

-- | Uniform scaling of a glyph.
newtype Scale = Scale RealN deriving (Eq, Show)

-- | Scaling of a glyph in the x-direction.
newtype ScaleX = ScaleX RealN deriving (Eq, Show)

-- | Scaling of a glyph in the y-direction.
newtype ScaleY = ScaleY RealN deriving (Eq, Show)

-- | Notes.
newtype Notes = Notes [Note] deriving (Eq, Show)

-- | Note.
data Note = Note Text [Color] deriving (Eq, Show)

-- | Multiple switches.
newtype Switches = Switches [Switch] deriving (Eq, Show)

-- | Switch list.
newtype Switch = Switch [SwitchArg] deriving (Eq, Show)

-- | Switch argument.
data SwitchArg
    = SwColor Color
    | SwShadeState ShadeState
    | SwSep Sep
    | SwFitState FitState
    | SwMirrorState MirrorState
    deriving (Eq, Show)

-- | Glyph separation.
newtype Sep = Sep RealN deriving (Eq, Show)

-- | Fit state.
data FitState
    = Fit
    | NoFit
    deriving (Eq, Show)

-- | Mirroring state.
data MirrorState
    = Mirror
    | NoMirror
    deriving (Eq, Show)

-- | Shade state.
data ShadeState
    = Shade
    | NoShade
    deriving (Eq, Show)

-- | A color.
data Color
    = Black
    | Red
    | Green
    | Blue
    | White
    | Aqua
    | Fuchsia
    | Gray
    | Lime
    | Maroon
    | Navy
    | Olive
    | Purple
    | Silver
    | Teal
    | Yellow
    deriving (Eq, Show)

-- | Area shading.
newtype ShadePattern = ShadePattern (NonEmpty Side)
    deriving (Eq, Show)

-- | Side to shade.
data Side
    = SideT
    | SideB
    | SideS
    | SideE
    deriving (Eq, Show)

-- | Real number, in the range 0.00 to 9.99.
--
-- Internally, the number is stored as a 'Word16' in hundredths precision
-- (ie. it has the range 0 to 999).
newtype RealN = RealN Word16 deriving (Eq, Ord)

instance Show RealN where
    show r = show ones <> "." <> show tenths <> show hundredths
        where (ones, tenths, hundredths) = realNDigits r

instance Num RealN where
    (RealN a) + (RealN b) =
        let c = a + b in if c <= 999 then RealN c else error "RealN overflow"
    (RealN a) - (RealN b)
        | b <= a    = RealN (b - a)
        | otherwise = error "RealN underfloat: there are no negative values"
    (*)    = error "multiplication is not yet implemented"
    negate = error
        "negate is not defined for RealN: there are no negative values"
    abs = id
    signum r@(RealN x) = if x == 0 then r else RealN 1
    fromInteger i = RealN (fromIntegral (i * 100))

-- | Creates an instance of a 'RealN' number, clamping digits as appropriate.
--
-- >>> mkRealN 0 5 2
-- 0.52
--
-- >>> mkRealN 5 2 5
-- 5.25
mkRealN
    :: Word8  -- ^ Ones digit (0-9).
    -> Word8  -- ^ Tenths digit (0-9).
    -> Word8  -- ^ Hundredths digit (0-9).
    -> RealN
mkRealN a b c = RealN (100 * f a + 10 * f b + f c)
  where
    f :: Word8 -> Word16
    f = fromIntegral . clamp 0 9

-- | Get the digits of a RealN number.
--
-- The digits are the ones, tenths and hundredths in a tuple. This is
-- effectively the inverse of 'mkReal'.
--
-- >>> realNDigits (mkRealN 0 4 2)
-- (0,4,2)
realNDigits :: RealN -> (Word8, Word8, Word8)
realNDigits (RealN w) = (ones, tenths, hundredths)
  where
    ones, tenths, hundredths :: Word8
    hundredths = f (w `mod` 10)
    tenths     = f ((w `div` 10) `mod` 10)
    ones       = f ((w `div` 100) `mod` 10)

    f :: Word16 -> Word8
    f = fromIntegral

-- | Clamp a value between two inclusive bounds.
--
-- >>> clamp 1.0 2.0 0.5
-- 1.0
--
-- >>> clamp 1.0 2.0 2.5
-- 2.0
--
-- >>> clamp 1.0 2.0 1.5
-- 1.5
clamp
    :: (Ord a)
    => a  -- ^ Minimum allowed value (inclusive).
    -> a  -- ^ Maximum allowed value (inclusive).
    -> a  -- ^ Input value.
    -> a  -- ^ Value clamped to lie between minimum and maximum inclusive.
clamp minAllowed maxAllowed x | x < minAllowed = minAllowed
                              | x > maxAllowed = maxAllowed
                              | otherwise      = x
